lenskit-core/src/main/java/org/lenskit/data/entities/Entity.java

```diff
@@ -0,0 +1,154 @@
+ * type, a numeric identifier (as a `long`), and additional data in a key-value mapping.
+ *
+ * Two entities are equal if their types, IDs, and data are equal.
+ *
+ * Identifiers are scoped per-type; it is acceptable to LensKit for two entities with different types to have the same
+ * identifier, and they are not considered to be the same entity.  However, individual data storage facilities are free
+ * to require identifiers to be disjoint.
+ */
+@Immutable
+@BuiltBy(BasicEntityBuilder.class)
+public interface Entity {
+    /**
+     * Get the identifier of this entity.
+     *
+     * @return The entity identifier.
+     */
+    long getId();
+
+    /**
+     * Get the type of this entity.
+     *
+     * @return The entity's type.
+     */
+    EntityType getType();
+
+    /**
+     * Get the names of the fields in this entity.
+     */
+    Set<String> getAttributeNames();
+
+    /**
+     * Get the fields in this entity.
+     */
+    Set<Attribute<?>> getAttributes();
+
+    /**
+     * Get the attribtue-value pairs.
+     */
+    Collection<AttributeValue<?>> getAttributeValues();
+
+    /**
+     * View this entity as a map.
+     * @return A map reflecting the entity's data.
+     */
+    Map<String,Object> asMap();
+
+    /**
+     * Check if the entity has a field with a particular name.
+     *
+     * @param name The field name to look for.
+     * @return `true` if the entity has a field named `name`.
+     */
+    boolean hasAttribute(String name);
+
+    /**
+     * Check if the entity has a attribute.
+     *
+     * @param attribute The attribute to look for.
+     * @return `true` if the entity contains `attribute` (same name **and type**).
+     */
+    boolean hasAttribute(Attribute<?> attribute);
+
+    /**
+     * Get the value of an attribute.
+     * @param attribute The attribute.
+     * @param <T> The attribute's type.
+     * @return The attribute's value.
+     * @throws NoSuchAttributeException if the specified attribute is not present.
+     * @throws IllegalArgumentException if a attribute with the same name as `attribute` is present, but it is of an
+     *         incompatible type.
+     */
+    @Nonnull
+    <T> T get(Attribute<T> attribute);
+
+    /**
+     * Get the value of an attribute by name.
+     * @param attr The attribute name.
+     * @throws NoSuchAttributeException if the specified attribute is not present.
+     * @return The field value, or `null` if the attribute is missing.
+     */
+    @Nonnull
+    Object get(String attr);
+
+    /**
+     * Get the value of a possibly-missing attribute.
+     * @param attribute The attribute.
+     * @param <T> The attribute's type.
+     * @return The attribute's value, or `null` if it is not present.
+     * @throws IllegalArgumentException if a attribute with the same name as `attribute` is present, but it is of an
+     *         incompatible type.
+     */
+    @Nullable
+    <T> T maybeGet(Attribute<T> attribute);
```

Comment:
Generics with classes confuse me. therefore I cannot predict what would happen if 
 1. You had Attribute<Class> and Attribute<SubClass> or Attribute<SuperClass> in these contexts.
 2. You have Attribute<Class> and an instance of SubClass